drivers/pwm/pwm_rcar.c

```diff
@@ -0,0 +1,279 @@
+	int ret = 0;
+
+	/* Assume only one channel per pwm instance */
+	if (pwm != 0) {
+		return -ENOTSUP;
+	}
+
+	if (flags != PWM_POLARITY_NORMAL) {
+		return -ENOTSUP;
+	}
+
+	/* Prohibiten values */
+	if (period_cycles == 0U || pulse_cycles == 0U ||
+	    pulse_cycles > period_cycles) {
+		return -EINVAL;
+	}
+
+	LOG_DBG("base_reg=0x%x, pulse_cycles=%d, period_cycles=%d,"
+		" duty_cycle=%d", config->reg_addr, pulse_cycles,
+		period_cycles, (pulse_cycles * 100U / period_cycles));
+
+	/* Disable PWM */
+	pwm_rcar_write_bit(config, RCAR_PWM_CR, RCAR_PWM_CR_EN, false);
+
+	/* Set continuous mode */
+	pwm_rcar_write_bit(config, RCAR_PWM_CR, RCAR_PWM_CR_SS, false);
+
+	/* Enable SYNC mode */
+	pwm_rcar_write_bit(config, RCAR_PWM_CR, RCAR_PWM_CR_SYNC, true);
+
+	/*
+	 * Set clock counter according to the requested period_cycles
+	 * if period_cycles is less than half of the counter, then the
+	 * clock diviser could be updated as the diviser is a modulo 2.
+	 */
+	if (period_cycles > RCAR_PWM_MAX_CYCLE ||
+	    period_cycles < (RCAR_PWM_MAX_CYCLE / 2)) {
+		LOG_DBG("Adapting frequency diviser...");
+		ret = pwm_rcar_update_clk(config, pwm, &period_cycles, &pulse_cycles);
+		if (ret != 0) {
+			return ret;
+		}
+	}
+
+	/* Set total period cycle */
+	reg_val = pwm_rcar_read(config, RCAR_PWM_CNT);
+	reg_val &= ~(RCAR_PWM_CNT_CYC_MASK);
+	reg_val |= (period_cycles << RCAR_PWM_CNT_CYC_SHIFT);
+	pwm_rcar_write(config, RCAR_PWM_CNT, reg_val);
+
+	/* Set high level period cycle */
+	reg_val = pwm_rcar_read(config, RCAR_PWM_CNT);
+	reg_val &= ~(RCAR_PWM_CNT_PH_MASK);
+	reg_val |= (pulse_cycles << RCAR_PWM_CNT_PH_SHIFT);
+	pwm_rcar_write(config, RCAR_PWM_CNT, reg_val);
+
+	/* Enable PWM */
+	pwm_rcar_write_bit(config, RCAR_PWM_CR, RCAR_PWM_CR_EN, true);
+
+	return ret;
+}
+
+static int pwm_rcar_get_cycles_per_sec(const struct device *dev,
+				       uint32_t pwm,
+				       uint64_t *cycles)
+{
+	const struct pwm_rcar_cfg *config = dev->config;
+	struct pwm_rcar_data *data = dev->data;
+	uint32_t diviser;
+
+	if (pwm != 0) {
+		return -ENOTSUP;
+	}
+
+	diviser = pwm_rcar_read(config, RCAR_PWM_CR) & RCAR_PWM_DIVISER_MASK;
+	diviser = diviser >> RCAR_PWM_DIVISER_SHIFT;
+	*cycles = data->clk_rate >> diviser;
+
+	LOG_DBG("Actual division: %d and Frequency: %d Hz", diviser, (uint32_t) *cycles);
+
+	return 0;
+}
+
+static int pwm_rcar_init(const struct device *dev)
+{
+	const struct pwm_rcar_cfg *config = dev->config;
+	struct pwm_rcar_data *data = dev->data;
+	int ret;
+
+	/* Configure dt provided device signals when available */
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = clock_control_on(config->clock_dev,
+			       (clock_control_subsys_t *)&config->mod_clk);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = clock_control_get_rate(config->clock_dev,
+				     (clock_control_subsys_t *)&config->core_clk,
+				     &data->clk_rate);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
```

Comment:
`int ret = 0;` then `return ret;` ?

Reply:
ret is always modified so no need to init it to 0.
I'm ok to simplify the end as `clock_control_get_rate` return 0 if no error.